app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquests/SpatialCache.kt

```diff
@@ -0,0 +1,140 @@
+package de.westnordost.streetcomplete.data.osm.osmquests
+
+import android.util.Log
+import de.westnordost.streetcomplete.data.download.tiles.TilePos
+import de.westnordost.streetcomplete.data.download.tiles.enclosingTilePos
+import de.westnordost.streetcomplete.data.download.tiles.enclosingTilesRect
+import de.westnordost.streetcomplete.data.download.tiles.minTileRect
+import de.westnordost.streetcomplete.data.osm.mapdata.BoundingBox
+import de.westnordost.streetcomplete.data.osm.mapdata.LatLon
+import de.westnordost.streetcomplete.util.math.contains
+import de.westnordost.streetcomplete.util.math.isCompletelyInside
+
+// inspired by Android LruCache
+// fetch needs get data from db and fill other caches, like questKey -> Quest
+// onRemoveTile needs to clean other caches, like questKey -> Quest
+// careful: put does not actually put key into cache if tile is not cached -> update db blocking if key was not put
+// TODO: synchronized stuff? didn't get related crash so far...
+// TODO: tile zoom as parameter? could smaller tiles help for mapDataController?
+class SpatialCache<T>(
+    private val maxSize: Int,
+    private val fetch: (BoundingBox) -> Map<T, LatLon>,
+    private val onEntriesRemoved: (Set<T>) -> Unit
+) {
+    private val byTile = LinkedHashMap<TilePos, HashSet<T>>((maxSize/0.75).toInt(), 0.75f, true)
+    private val byKey = hashMapOf<T, LatLon>()
+
+    fun remove(key: T) {
+        val pos = byKey.remove(key) ?: return
+        byTile[getTilePosFor(pos)]?.remove(key)
+        onEntriesRemoved(setOf(key))
+    }
+
+    fun removeAll(keys: Collection<T>) {
+        val removed = keys.filter {
+            val pos = byKey.remove(it) ?: return@filter false
+            byTile[getTilePosFor(pos)]?.remove(it)
+            true
+        }
+        onEntriesRemoved(removed.toSet())
+    }
+
+    // returns whether the key was put in cache
+    fun putIfTileExists(key: T, position: LatLon): Boolean {
+        val previousPosition = byKey[key]
+        // in cache already
+        if (previousPosition != null) {
+            // but moved -> remove
+            if (previousPosition != position) remove(key)
+            else return true
+        }
+        val tileInCache = byTile[getTilePosFor(position)]
+        if (tileInCache  != null) {
+            tileInCache.add(key)
+            byKey[key] = position
+            return true
+        }
+        return false
+    }
+
+    fun replaceAllInBBox(entries: Map<T, LatLon>, bbox: BoundingBox) {
+        replaceAllInTiles(entries, bbox.asListOfEnclosingTilePos())
```

Comment:
So in reality / in our system, the tiles will probably always align with the bboxes. Still, there should be a sanity check here. In case the tiles and bbox do not align, what should be done instead is to remove the affected tiles from cache. (Add a log for this situation because in case this is called (too often) due to floating point imprecision, we need to do something about that)

Reply:
Right, looks like I forgot this when moving from the initial commit to the `SpatialCache`

Reply:
done